arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp

```diff
@@ -98,54 +97,77 @@ auto DocumentLeaderState::recoverEntries(std::unique_ptr<EntryIterator> ptr)
       return TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED;
     }
 
-    auto transactionHandler = self->_handlersFactory->createTransactionHandler(
-        data.core->getVocbase(), self->gid);
+    auto transactionHandler = data.core->getTransactionHandler();
     ADB_PROD_ASSERT(transactionHandler != nullptr);
 
+    std::unordered_set<TransactionId> activeTransactions;
     while (auto entry = ptr->next()) {
       auto doc = entry->second;
-      if (doc.operation == OperationType::kCreateShard) {
-        auto const& collectionId = doc.collectionId;
-        auto res = data.core->ensureShard(doc.shardId, collectionId, doc.data);
-        if (res.fail()) {
-          LOG_CTX("9e993", FATAL, self->loggerContext)
-              << "Failed to create shard " << doc.shardId << " of collection "
-              << collectionId << " with error: " << res;
-          FATAL_ERROR_EXIT();
-        }
-      } else if (doc.operation == OperationType::kDropShard) {
-        auto const& collectionId = doc.collectionId;
-        transactionHandler->abortTransactionsForShard(doc.shardId);
-        auto res = data.core->dropShard(doc.shardId, collectionId);
-        if (res.fail()) {
-          LOG_CTX("f1eb0", FATAL, self->loggerContext)
-              << "Failed to drop shard " << doc.shardId << " of collection "
-              << collectionId << " with error: " << res;
-          FATAL_ERROR_EXIT();
-        }
-      } else {
-        // Note that the shard could've been dropped before doing recovery.
-        if (!data.core->isShardAvailable(doc.shardId)) {
-          LOG_CTX("bee57", INFO, self->loggerContext)
-              << "Will not apply transaction " << doc.tid << " for shard "
-              << doc.shardId << " because it is not available";
-          continue;
-        }
 
-        auto res = transactionHandler->applyEntry(doc);
-        if (res.fail()) {
-          return res;
-        }
+      auto res = std::visit(
+          [&](auto&& op) -> Result {
+            using T = std::decay_t<decltype(op)>;
+            if constexpr (ModifiesUserTransaction<T>) {
```

Comment:
The problem is a bit more complicated for `recoverEntries`, because there are several places where we return early from the `std::visit` call. There is also some common code at the end, the `transactionHandler->applyEntry` call, which might happen or not, depending on the logic contained in the if statements above.